### DISCUSSION TITLE
yml配置中${NACOS_IP:@nacos.ip@} 是什么意思

### DISCUSSION THREAD

**Discussion by zuihou** [zuihou/zuihou-admin-cloud#58] [02.提问]
# 版本信息：
JDK 版本(必填) : 1.8.0
源码版本(必填)：所有版本

# 问题描述：
yml配置中 `${NACOS_IP:@nacos.ip@}` 是什么意思?

# 报错截图
![image](https://user-images.githubusercontent.com/8500410/94144175-65d7fe80-fea3-11ea-9b76-bc91ec380423.png)

## 支持一下:
如果该issue帮助了你, 麻烦给项目点个star, 并给当前issue回复一下 “👍” 或者 “已解决”.

**Comment by zuihou** [comment]
1. `${NACOS_IP}` 表示从环境变量、yml配置上下文、外部配置文件、命令行参数等去查找 参数 NACOS_IP 
2. `:` 表示 NACOS_IP在第1条中说的地方找不到值时, 使用 `:` 后面的默认值
3. `@nacos.ip@` 表示在使用maven编译期间,会从maven上下文 查找 nacos.ip .
![image](https://user-images.githubusercontent.com/8500410/189598345-d294487d-9b46-4a1a-a1db-59eca7bac6c6.png)
4. `@project.*@`  以 project开头的参数，在使用mvn 命令编译项目时，maven 的pom文件中读取参数
![image](https://user-images.githubusercontent.com/8500410/189598170-653cca29-f016-4292-b0ab-e138e90c9983.png)
5. `@profile.active@` 会从profiles节点读取参数
```
    <profiles>
        <profile>
            <id>dev</id>
            <properties>
                <profile.active>dev</profile.active>
            </properties>
            <activation>
                <activeByDefault>true</activeByDefault>
            </activation>
        </profile>
        <profile>
            <id>prod</id>
            <properties>
                <profile.active>prod</profile.active>
            </properties>
        </profile>
    </profiles>
```
6. nacos.ip 的值可以通过pom文件、 src/main/filters/config-dev.properties 文件传递过去
7. `@` 符号在那里定义的?  lamp-cloud/pom.xml 
```
                 <plugin>
                    <groupId>org.apache.maven.plugins</groupId>
                    <artifactId>maven-resources-plugin</artifactId>
                    <version>${maven-resources-plugin.version}</version>
                    <configuration>
                        <delimiters>
                            <delimiter>@</delimiter>
                        </delimiters>
                    </configuration>
                </plugin>
```
9. @@  变量不仅仅能在yml文件中使用, 还支持在src/main/resources文件夹下的所有文本文件中使用. 如: banner.txt
10. 在编译期，编译期间执行 `mvn clean  install -P prod` ,  表示编译时加载config-prod.properties 文件.
11. 在编译期，编译期间执行 `mvn clean  install -P dev` ,  表示编译时加载config-dev.properties 文件.

**Reply by zqb666** [reply]
name: @project.artifactId@ 比如这种呢？ 是怎样的读取逻辑

**Reply by zuihou** [reply]
project.xxx 等是读取的 pom.xml 中的属性。

**Comment by zuihou** [comment]
在配置文件中使用@@变量的优缺点:
优点:
1. 可以将多个服务的 bootstrap.yml、banner.txt 等配置文件中相同的配置抽离出去,达到全局配置的效果.

缺点:
1. 不支持独立服务单独编译.
 如: 使用jenkins自动构建时, 想要单独部署权限服务等服务时,  若想单独编译权限服务,会报找不到config-dev.properties配置文件等错.  只能采取拉整个工程的代码来编译后, 进入到权限服务文件夹, 然后独立部署权限服务.
